### PR TITLE
8.0 stock quant merge

### PR DESCRIPTION
**Pull request by clonedagain**
Proposal to adopt OdooMRP's module to merge quants when they were split and it's not needed anymore (ie. reservation is undone).

To be done:
- [x] Add OCA as co-author
- [x] Use OCA's README template
- [x] Add automatic tests

### CODE REVIEW

**Comment by clonedagain** [comment]
For you information @pedrobaeza @oihane @anajuaristi @lbellier

**Comment by anajuaristi** [comment]
Stock_quant_merge is working in production for several customers so :+1: RTM from my side

**Comment by rvalyi** [comment]
Looks good. It's something I always wanted.
But @clonedagain remarks should be implemented so we can merge it in the OCA.

**Comment by clonedagain** [comment]
The code now adheres to the conventions and has automatic tests.
Reviewers wanted.

**Comment by elicoidal** [comment]
Formally LGTM (no test, no technical check)

File: stock_quant_merge/models/stock.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+class StockQuant(models.Model):
+    _inherit = 'stock.quant'
+
+    @api.multi
+    def merge_stock_quants(self):
+        pending_quants = self.filtered(lambda x: True)
+        for quant2merge in self:
+            if (quant2merge in pending_quants and
+                    not quant2merge.reservation_id):
+                quants = self.search(
```
Comment (rvalyi) [inline stock_quant_merge/models/stock.py]: could you please isolate this domain in a method that would take quant2merge as an argument and return the domain?
Some may need to override this domain. Example: in Brazil we will probably separate quants of the same product depending on if it comes from national origin or if it's imported. It's the same product in the MRP for instance. But for fiscal reports we cannot mix quants and we don't want to freak with lots for product origin. In this case we would need to override this domain, but may be there are other use cases.

File: stock_quant_merge/models/stock.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+from openerp import models, api
+
+
+class StockQuant(models.Model):
+    _inherit = 'stock.quant'
+
+    @api.multi
+    def merge_stock_quants(self):
+        pending_quants = self.filtered(lambda x: True)
```
Comment (oihane) [inline stock_quant_merge/models/stock.py]: Could you please explain why do you use pending_quants, wouldn't it been enough by processing those with `reservation_id` not set?

Comment (clonedagain) [inline stock_quant_merge/models/stock.py]: Wait, _you_ wrote that didn't you? :-)
I think it avoids trying to merge a quant we just removed in a previous iteration.

Comment (oihane) [inline stock_quant_merge/models/stock.py]: Yes, I did, but I don't remember why :worried: 
Maybe you should also start by doing the for only with those quants that do not have reservation...
I mean `for quant2merge in self.filtered(lambda x: not x.reservation_id)`

**Comment by clonedagain** [comment]
@rvalyi i've changed it as you suggested.

**Comment by rvalyi** [comment]
@clonedagain thanks! and kudos for the work!

:+1:

File: stock_quant_merge/models/stock.py (hunk 2 above)
Comment (clonedagain) [inline stock_quant_merge/models/stock.py]: I suppose that's a way to get a copy of the recordset?

File: stock_quant_merge/models/stock.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                ('lot_id', '=', self.lot_id.id),
+                ('package_id', '=', self.package_id.id),
+                ('location_id', '=', self.location_id.id),
+                ('reservation_id', '=', False),
+                ('propagated_from_id', '=', self.propagated_from_id.id)]
+
+    @api.multi
+    def merge_stock_quants(self):
+        # Get a copy of the recorset
+        pending_quants = self.browse(self.ids)
```
Comment (clonedagain) [inline stock_quant_merge/models/stock.py:28]: @oihane is it clearer this way?

Comment (oihane) [inline stock_quant_merge/models/stock.py:28]: why don't you change the _for_? If you get only the ones with reservation_id not set it will do less work, don't you think?

Comment (clonedagain) [inline stock_quant_merge/models/stock.py:28]: The `for` is on `self` not `pending_quants`.

Comment (clonedagain) [inline stock_quant_merge/models/stock.py:28]: For the main part I think it's OK because the module itself only calls this method on unreserved quants, but there may be corner cases if called through RPC or custom modules.

I'm sorry but it's a bit too subtle for me to research given the time I can spare on this.
Maybe you could propose a test to make sure it's OK?

Comment (oihane) [inline stock_quant_merge/models/stock.py:28]: This is what I mean...

```
    @api.multi
    def merge_stock_quants(self):
        # Get a copy of the recorset
        pending_quants = self.browse(self.ids)
        for quant2merge in self.filtered(lambda x: not x.reservation_id):
            if quant2merge in pending_quants:
                quants = self.search(self._mergeable_domain())
                cont = 1
                cost = quant2merge.cost
                for quant in quants:
                    if (self._get_latest_move(quant2merge) ==
                            self._get_latest_move(quant)):
                        quant2merge.sudo().qty += quant.qty
                        cost += quant.cost
                        cont += 1
                        pending_quants -= quant
                        quant.sudo().unlink()
                quant2merge.sudo().cost = cost / cont
```

So this way you will only **for** those quants that have not been reserved yet.

@pedrobaeza, am I wrong?

Comment (clonedagain) [inline stock_quant_merge/models/stock.py:28]: Ok changed as you suggested.

**Comment by clonedagain** [comment]
@lbellier is it OK for you?

**Comment by lbellier** [comment]
@clonedagain Yes, it's good for me.

**Comment by Cedric-Pigeon** [comment]
:+1: (code review) Thanks !

**Comment by oihane** [comment]
I see it ok :+1:
